### PR TITLE
feat(ast_tools): support `#[scope(exit_before)]`

### LINKED ISSUE #6311: Support exiting scope before last field `#[scope(exit_before)]`
`oxc_ast_tools` recognises `#[scope(enter_before)]` to mark a field as the start of the scope in a type. We also need `#[scope(exit_before)]` for #6270.

Note: I think `#[scope(exit_before)]` is better than `#[scope(exit_after)]` because attributes go before fields, so it reads more clearly:

```rs
pub struct TSConditionalType<'a> {
    pub span: Span,
    pub check_type: TSType<'a>,
    #[scope(enter_before)]
    pub extends_type: TSType<'a>,
    pub true_type: TSType<'a>,
    #[scope(exit_before)]
    pub false_type: TSType<'a>,
    pub scope_id: Cell<Option<ScopeId>>,
}
```

We will need to build support for this into both:

1. `oxc_ast_tools` and the `Visit` / `VisitMut` generator.
2. `oxc_traverse`'s codegen.

### PR DESCRIPTION
**Pull request by DonIsaac**
Closes #6311.

Adds support for `#[scope(exit_before)]`, which is the opposite of `#[scope(enter_before)]`

### CODE REVIEW

**Comment by DonIsaac** [comment]
* **#6350** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6350?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by Dunqing** [comment]
From the description of #6311, we should be supporting `#[scope(exit_before)]` rather than `#[scope(exit_after]`. Let me know if I got it wrong.

**Review by overlookmotel — CHANGES_REQUESTED** [review]
As mentioned in #6311, I think `#[scope(exit_before)]` is better than `#[scope(exit_after)]`, because then when you read the type def, the `#[scope(exit_before)]` attr is at the point where the scope is exited.

e.g.:

```rs
pub struct TSConditionalType<'a> {
    pub span: Span,
    pub check_type: TSType<'a>,
    #[scope(enter_before)]
    pub extends_type: TSType<'a>,
    pub true_type: TSType<'a>,
    #[scope(exit_before)]
    pub false_type: TSType<'a>,
    pub scope_id: Cell<Option<ScopeId>>,
}
```

The fields that are inside the scope are `extends_type` and `true_type`, and they are bounded by the `#[scope(...)]` attrs on each side.

Is there a reason why you decided against that?

**Review by overlookmotel — DISMISSED** [review]
`Traverse` codegen also needs to support the new exit attr.

https://github.com/oxc-project/oxc/tree/main/crates/oxc_traverse/scripts

You can see in #6351 that `walk.rs` is not exiting the scope when it should: https://github.com/oxc-project/oxc/pull/6351/files#diff-f70f8cd02df83ed234c9b27dcb7ea2c948b2ab6110952e87eadaf0ddafbd6b9a

File: tasks/ast_tools/src/markers.rs [hunk 1]
```diff
@@ -64,7 +64,10 @@ pub struct VisitMarkers {
 /// A struct representing `#[scope(...)]` markers
 #[derive(Default, Debug)]
 pub struct ScopeMarkers {
+    /// `#[scope(enter_before)]`
```
Comment (Boshen) [inline tasks/ast_tools/src/markers.rs:67]: Unintentional commented out code?

Comment (overlookmotel) [reply tasks/ast_tools/src/markers.rs:67]: I believe this is intentional. It's a doc comment explaining what attribute `ScopeMarkers::enter_before` represents.

**Comment by DonIsaac** [comment]
> `Traverse` codegen also needs to support the new exit attr.
> 
> https://github.com/oxc-project/oxc/tree/main/crates/oxc_traverse/scripts
> 
> You can see in #6351 that `walk.rs` is not exiting the scope when it should: https://github.com/oxc-project/oxc/pull/6351/files#diff-f70f8cd02df83ed234c9b27dcb7ea2c948b2ab6110952e87eadaf0ddafbd6b9a

Fixed. Ready for another review.

File: crates/oxc_traverse/scripts/lib/walk.mjs [hunk 2]
```diff
@@ -92,6 +94,20 @@ function generateWalkForStruct(type, types) {
...
+      exitScopeField = visitedFields.find(field => field.name === exitFieldName);
+      assert(
+        exitScopeField,
+        `\`ast\` attr says to exit scope before field '${exitFieldName}' ` +
+          `in '${type.name}', but that field is not visited`,
+      );
+    }
+    // else {
+    //   exitScopeField = visitedFields[visitedFields.length - 1];
+    // }
```
Comment (overlookmotel) [inline crates/oxc_traverse/scripts/lib/walk.mjs:109]: Remove commented-out code.

```suggestion
```

File: crates/oxc_traverse/scripts/lib/walk.mjs [hunk 3]
```diff
@@ -75,7 +75,9 @@ function generateWalkForStruct(type, types) {
 
   const { scopeArgs } = type;
   /** @type {Field | undefined} */
-  let scopeEnterField;
+  let scopeEnterField,
+    /** @type {Field | undefined} */
+    exitScopeField;
```
Comment (overlookmotel) [inline crates/oxc_traverse/scripts/lib/walk.mjs:80]: Could you name this var `scopeExitField` for consistency with `scopeEnterField`?

File: tasks/ast_tools/src/generators/visit.rs [hunk 4]
```diff
@@ -525,6 +527,18 @@ impl<'a> VisitBuilder<'a> {
...
+                if have_exit_scope {
+                    assert!(
+                        exit_scope_at.is_none(),
+                        "Scopes cannot be exited more than once. Remove the extra `#[scope(exit_before)]` attribute(s)."
+                    );
+                    let scope_leave = &scope_events.1;
+                    result = quote! {
+                        #scope_leave
+                        #result
+                    };
```
Comment (overlookmotel) [inline tasks/ast_tools/src/generators/visit.rs:539]: Rename var for consistency.

```suggestion
                    let scope_exit = &scope_events.1;
                    result = quote! {
                        #scope_exit
                        #result
                    };
```

**Review by overlookmotel — COMMENTED** [review]
Sorry for the nits...

**Pull request merged**